presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RewriteExpressionWithConstantVariable.java

```diff
@@ -0,0 +1,391 @@
+                newFilterNode = new FilterNode(node.getSourceLocation(), idAllocator.getNextId(), node.getSource(), predicate);
+                outputConstantMap.putAll(rewrittenChild.getConstantExpressionMap());
+            }
+
+            Map<VariableReferenceExpression, ConstantExpression> newConstantMap = new HashMap<>();
+            for (RowExpression conjunct : extractConjuncts(predicate)) {
+                if (conjunct instanceof CallExpression && ((CallExpression) conjunct).getFunctionHandle().getName().equals(EQUAL.getFunctionName().toString())) {
+                    RowExpression argument0 = ((CallExpression) conjunct).getArguments().get(0);
+                    RowExpression argument1 = ((CallExpression) conjunct).getArguments().get(1);
+                    if (isSupportedType(argument0) && isSupportedType(argument1)) {
+                        if (argument0 instanceof VariableReferenceExpression && argument1 instanceof ConstantExpression) {
+                            if (newConstantMap.containsKey(argument0) && !newConstantMap.get(argument0).equals(argument1)) {
+                                return new PlanNodeWithConstant(newFilterNode.replaceChildren(ImmutableList.of(rewrittenChild.getPlanNode())), ImmutableMap.of());
+                            }
+                            newConstantMap.put((VariableReferenceExpression) argument0, (ConstantExpression) argument1);
+                        }
+                        else if (argument1 instanceof VariableReferenceExpression && argument0 instanceof ConstantExpression) {
+                            if (newConstantMap.containsKey(argument1) && !newConstantMap.get(argument0).equals(argument0)) {
+                                return new PlanNodeWithConstant(newFilterNode.replaceChildren(ImmutableList.of(rewrittenChild.getPlanNode())), ImmutableMap.of());
+                            }
+                            newConstantMap.put((VariableReferenceExpression) argument1, (ConstantExpression) argument0);
+                        }
+                    }
+                }
+            }
+            outputConstantMap.putAll(newConstantMap);
+            return new PlanNodeWithConstant(newFilterNode.replaceChildren(ImmutableList.of(rewrittenChild.getPlanNode())), outputConstantMap.build());
+        }
+
+        @Override
+        public PlanNodeWithConstant visitProject(ProjectNode node, Void context)
+        {
+            PlanNodeWithConstant rewrittenChild = accept(node.getSource());
+            ImmutableMap.Builder<VariableReferenceExpression, ConstantExpression> outputConstantMap = ImmutableMap.builder();
+            ProjectNode newProjectNode = node;
+            if (!rewrittenChild.getConstantExpressionMap().isEmpty()) {
+                Map<VariableReferenceExpression, RowExpression> newAssignments = node.getAssignments().getMap().entrySet().stream()
```

Comment:
what happens if the constant is not of the same type as the variable? will this change the type of the assignment?

Reply:
The variable and constant are populated from the assignments in projection and predicate in filter, should already be of the same type. I also added a check statement to throw exceptions if this does not hold.

Reply:
do we always introduce casts if there is a type mismatch (and it's a supported pair of types)? and if yes, I assume that's been done before this optimization?

Reply:
My understanding is that it's done during query plan generation before optimization https://github.com/prestodb/presto/blob/650ca5de0b443df499f66b7b43ec31538bd8e2ce/presto-main/src/main/java/com/facebook/presto/sql/planner/TranslationMap.java#L272

Reply:
we add casts for all type mismatches during planning, and then in RowExpressionInterpreter remove casts for type-only coercions.  Looks like that gets called from LocalExecutionPlanner, which is later than I expected. I wonder if it used to be earlier before migrating everything to RowExpressions.